Increase TOTP coverage

iandunn opened the pull request:
This adds some tests to fill small gaps in coverage, and marks some functions as ignored because they're not testable in a unit test, or aren't worth testing. This increases the coverage of `class-two-factor-totp.php` to `92%`, making it obvious that it is well-tested.

To aid in writing tests, this also adds `phpunit-watcher`, similar to Gutenberg, WordCamp.org, and some other WordPress repos.

See #468

iandunn commented:
I'll take a closer look tomorrow at the lint-php8 failure tomorrow.

kasparsd commented:
The PHP8 linter issues are fixed in #466.

kasparsd commented on composer.json:
```diff
@@ -11,8 +11,9 @@
   },
   "config": {
     "sort-packages": true,
+    "process-timeout": 0,
```
This is set to 300 seconds or 5 minutes by default. I think we should keep it at that in order to avoid long running CI tasks. All cases with linters or tests running for more than 5 minutes should probably be investigated and made to run faster. Keeping the limit in place will ensure that we notice when this happens.

iandunn replied:
In the past it's caused problems with the `watch` task, since it's designed to run for hours, but I'll test to make sure that's still the case, and if there's another way around it.

kasparsd replied:
I was able to fix it like so 466370bc928fffa71e6d80f0009cb0ee0ce75bb2. Does that solve it?

kasparsd commented on readme.md:
```diff
@@ -14,12 +14,18 @@ Please [report issues](https://github.com/WordPress/two-factor/issues) and [open
 Join the `#core-passwords` channel [on WordPress Slack](http://wordpress.slack.com) ([sign up here](http://chat.wordpress.org)).
 
-Here is how to get started:
+To use the provided development environment, you'll first need to install and launch Docker. Once it's running, the next steps are:
```
Thanks for updating these instructions!

kasparsd commented on composer.json:
```diff
@@ -11,8 +11,9 @@
     "sort-packages": true,
+    "process-timeout": 0,
     "platform": {
-      "php": "5.6.20"
```
Maybe leave this to a separate pull request as we'll probably want to bump the versions of phpunit when doing this. The `phpunit-polyfills` are already present but we'll need to update the actual test files to match the new parent class signatures for the phpunit base classes. 

Overall, we could have this platform locked at the version we're using/expecting for all the dev tooling while the phpcompatability checker is used for ensuring that all code is compatible with the lowest version of PHP supported by the WP core, right?

kasparsd commented on composer.json:
```diff
@@ -44,6 +45,7 @@
     "lint": "phpcs",
     "lint-php8": "phpcs -p --standard=PHPCompatibilityWP --runtime-set testVersion 8.0- --extensions=php --ignore='vendor/,wordpress/,node_modules/' .",
     "test": "phpunit",
+    "test:watch": "phpunit-watcher watch < /dev/tty",
```
It seems that `< /dev/tty` might not be necessary. I was able to start the watcher without it and it worked fine.

kasparsd commented on package.json:
```diff
@@ -19,8 +19,8 @@
-    "test": "npm-run-all test:*",
-    "test:php": "npm run composer test",
+    "test": "npm run composer test",
+    "test:watch": "npm run composer test:watch -- -- --timeout=0",
```
While resolving the merge conflict, I was able to move the timeout logic here where I assume it was needed. Please correct me if I'm wrong.

iandunn replied:
Ah that's a great idea, I'll test that out 👍🏻

kasparsd commented on phpunit-watcher.yml.dist:
```diff
@@ -0,0 +1,9 @@
+watch:
+  directories:
+    - ./providers/
```
Should we add the root directory to this list to ensure that changes to those files also trigger the runs?

iandunn replied:
We could, as long as we exclude `vendor` etc. I'll look into that 👍🏻

kasparsd commented on .github/workflows/test.yml:
```diff
@@ -20,7 +20,7 @@ jobs:
         uses: shivammathur/setup-php@v2
         with:
-          php-version: '5.6'
+          php-version: '7.2'
```
The php8 compat packages need a later version of PHP. All the linters are running correctly now.

iandunn commented:
I removed the commits related to `phpunit-watcher` and rebased against `master`, so I think this is ready for review again. I'll open a new PR w/ the other commits and do some of the other PHP tooling work there.

kasparsd reviewed (approved):
Looks great!

kasparsd commented:
Yes, please merge it in!
